File: 2024/clojure/src/day04.clj
(ns day04
  (:require
   [clojure.string :as str]))

(def directions
  [[0 1]
   [1 0]
   [1 1]
   [1 -1]
   [0 -1]
   [-1 0]
   [-1 -1]
   [-1 1]])

(defn in-bounds? [grid x y]
  (and (>= x 0)
       (< x (count grid))
       (>= y 0)
       (< y (count (first grid)))))

(defn get-letter [grid x y]
  (when (in-bounds? grid x y)
    (get-in grid [x y])))

(defn search-xmas [grid x y dx dy]
  (let [word "XMAS"
        letters (map #(get-letter grid (+ x (* dx %)) (+ y (* dy %))) (range (count word)))]
    (if (= (seq word) letters) 1 0)))

(defn count-xmas [grid]
  (reduce
   (fn [total [x y]]
     (+ total
        (reduce
         (fn [dir-total [dx dy]]
           (+ dir-total (search-xmas grid x y dx dy)))
         0
         directions)))
   0
   (for [x (range (count grid))
         y (range (count (first grid)))]
     [x y])))

(defn check-x-mas [grid x y]
  (let [diag1 (map #(get-letter grid (+ x %) (+ y %)) [0 1 2])
        diag2 (map #(get-letter grid (- (+ x 2) %) (+ y %)) [0 1 2])]
    (if (and
         (or (= diag1 (seq "MAS")) (= diag1 (seq "SAM")))
         (or (= diag2 (seq "MAS")) (= diag2 (seq "SAM"))))
      1
      0)))

(defn count-x-mas [grid]
  (reduce
   (fn [total [x y]]
     (+ total (check-x-mas grid x y)))
   0
   (for [x (range (count grid))
         y (range (count (first grid)))]
     [x y])))

(defn parse-input [input]
  (mapv vec (str/split-lines input)))

(let [input "MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX"
      parsed-input (parse-input input)]
  (println "Part 1 (Example):" (count-xmas parsed-input))
  (println "Part 2 (Example):" (count-x-mas parsed-input)))

(let [input (slurp "data/day04.txt")
      parsed-input (parse-input input)]
  (println "Part 1:" (count-xmas parsed-input))
  (println "Part 1:" (count-x-mas parsed-input)))

File: 2024/python/day04.py
import argparse

from utils import read_lines

parser = argparse.ArgumentParser()
parser.add_argument("input", type=str, help="Path to input file")


directions = [
    [0, 1],
    [1, 0],
    [1, 1],
    [1, -1],
    [0, -1],
    [-1, 0],
    [-1, -1],
    [-1, 1],
]


def in_bounds(grid, x, y):
    return 0 <= x < len(grid) and 0 <= y < len(grid[0])


def get_letter(grid, x, y):
    if in_bounds(grid, x, y):
        return grid[x][y]


def search_xmas(grid, x, y, dx, dy):
    word = list("XMAS")
    letters = [get_letter(grid, x + (dx * i), y + (dy * i)) for i in range(len(word))]
    return letters == word


def count_xmas(grid):
    coords = [[x, y] for x in range(len(grid)) for y in range(len(grid[0]))]

    total = 0

    for x, y in coords:
        for dx, dy in directions:
            total += search_xmas(grid, x, y, dx, dy)

    return total


def part_1(x):
    return count_xmas(x)


def check_x_mas(grid, x, y):
    diag1 = [get_letter(grid, x + d, y + d) for d in [0, 1, 2]]
    diag2 = [get_letter(grid, x + 2 - d, y + d) for d in [0, 1, 2]]

    return (diag1 == list("MAS") or diag1 == list("SAM")) and (
        diag2 == list("MAS") or diag2 == list("SAM")
    )


def count_x_mas(grid):
    coords = [[x, y] for x in range(len(grid)) for y in range(len(grid[0]))]

    total = 0

    for x, y in coords:
        total += check_x_mas(grid, x, y)

    return total


def part_2(x):
    return count_x_mas(x)


def main():
    args = parser.parse_args()
    x = read_lines(args.input)

    print("Part 1:", part_1(x))
    print("Part 2:", part_2(x))


if __name__ == "__main__":
    main()
